packages/jest-cli/src/watch.js

```diff
@@ -18,23 +17,54 @@ import exit from 'exit';
 import {replacePathSepForRegex} from 'jest-regex-util';
 import HasteMap from 'jest-haste-map';
 import isValidPath from './lib/is_valid_path';
-import {getFailedSnapshotTests, isInteractive} from 'jest-util';
+import {isInteractive} from 'jest-util';
 import {print as preRunMessagePrint} from './pre_run_message';
 import createContext from './lib/create_context';
 import runJest from './run_jest';
 import updateGlobalConfig from './lib/update_global_config';
 import SearchSource from './search_source';
-import SnapshotInteractiveMode from './snapshot_interactive_mode';
 import TestWatcher from './test_watcher';
 import Prompt from './lib/Prompt';
-import TestPathPatternPrompt from './test_path_pattern_prompt';
-import TestNamePatternPrompt from './test_name_pattern_prompt';
 import FailedTestsCache from './failed_tests_cache';
-import WatchPluginRegistry from './lib/watch_plugin_registry';
 import {KEYS, CLEAR} from './constants';
+import JestHooks from './jest_hooks';
+import WatchPlugin from './watch_plugin';
+import TestPathPatternPlugin from './plugins/test_path_pattern';
+import TestNamePatternPlugin from './plugins/test_name_pattern';
+import UpdateSnapshotsPlugin from './plugins/update_snapshots';
+import UpdateSnapshotsInteractivePlugin from './plugins/update_snapshots_interactive';
+import QuitPlugin from './plugins/quit';
 
 let hasExitListener = false;
 
+const INTERNAL_PLUGINS = [
+  TestPathPatternPlugin,
+  TestNamePatternPlugin,
+  UpdateSnapshotsPlugin,
+  UpdateSnapshotsInteractivePlugin,
+  QuitPlugin,
+];
+
+const getSortedUsageRows = (
+  watchPlugins: Array<WatchPlugin>,
+  globalConfig: GlobalConfig,
+  hasSnapshotFailure: boolean,
+) => {
+  const internalPlugins = watchPlugins
+    .slice(0, INTERNAL_PLUGINS.length)
+    .map(p => p.getUsageRow(globalConfig, hasSnapshotFailure))
+    .filter(usage => usage && !usage.hide);
+
+  const thirdPartyPlugins = watchPlugins
+    .slice(INTERNAL_PLUGINS.length)
+    .map(p => p.getUsageRow(globalConfig, hasSnapshotFailure))
```

Comment:
`hasSnapshotFailure` seems really specific, IDK if there's a better place to stick it, though

Reply:
`testResults` array would maybe make more sense, possible with some properties like `hasSnapshotFailure: boolean` on it for ease of use

Reply:
Agree!

This is something that I was thinking over the weekend, I'm glad that you felt the same way!

My thoughts are that `getUsageRow()` should only take in the `globalFilter`.

Stuff like `hasSnapshotFailure` (which as you mentioned only a small chunck of plugins will use) can be grabbed from the`testRunComplete` hook.


```js
hooks.testRunComplete(results => {
  this._hasSnapshotFailure = !!results.snapshot.failure;
});
```

### Unrelated comment

Which makes me think that we could take this approach even further and not even pass `globalConfig` as a param of `getUsageRow`, but instead if the plugin need access to the most up to date `globalConfig` it can get it from a `globalConfigChanged` hook

```js
hooks.globalConfigChanged(globalConfig => {
  this._globalConfig = globalConfig;
});
```


What are your thoughts on this?